text/0000-assoc-default-groups.md

```diff
@@ -698,6 +739,199 @@ to change this behaviour. Presumably, this is the case because Haskell
+
+Since Idris is language with full spectrum dependent types,
+it does not distinguish between terms and types, instead, types are terms.
+Therefore, there is really not a distinct concept called "associated type".
+However, an `interface` may require certain definitions to be provided
+and this includes types. For example, we may write:
+
+```idris
+interface Iterator self where
+    item : Type
+    next : self -> Maybe (self, item)
+
+implementation Iterator (List a) where
+    item = a
+    next [] = Nothing
+    next (x :: xs) = Just (xs, x)
+```
+
+Like in Haskell, in Idris, a function or value in an interface may be given a
+default definition. For example, the following is a valid program:
+
+```idris
+interface Foo x where
+    bar : Type
+    bar = Bool
+
+    baz : x -> bar
+
+implementation Foo Int where
+    baz x = x == 0
+```
+
+However, if we provide a default for `baz` in the `interface` which assumes
+the default value `Bool` of `bar`, as with the following example:
+
+```idris
+interface Foo x where
+    bar : Type
+    bar = Bool
+
+    baz : x -> bar
+    baz _ = True
+```
+
+then we run into an error:
+
+```
+Type checking .\foo.idr
+foo.idr:6:13-16:
+  |
+6 |     baz _ = True
+  |             ~~~~
+When checking right hand side of Main.default#baz with expected type
+        bar x _
+
+Type mismatch between
+        Bool (Type of True)
+and
+        bar x _ (Expected type)
+```
+
+The behaviour here is exactly as in Haskell and as proposed in this RFC.
+
+## C++
+
+In C++, it is possible to provide associated types and specialize them as well.
+This is shown in the following example:
+
+```cpp
+#include <iostream>
+#include <string>
+
+template<typename T> struct wrap {};
+
+template<typename T> struct foo { // Unspecialized.
+    typedef int bar;
+
+    bar make_a_bar() { return 0; };
+};
+
+template<typename T> struct foo<wrap<T>> { // Partial specialization.
+    typedef std::string bar;
+
+    bar make_a_bar() { return std::string("hello world"); };
+};
+
+int main() {
+    foo<void> a_foo;
+    std::cout << a_foo.make_a_bar() << std::endl;
+
+    foo<wrap<void>> b_foo;
+    std::cout << b_foo.make_a_bar() << std::endl;
+}
+```
+
+One thing to note here is that C++ allows us to assume in both the unspecialized
+variant of `foo` as well as the specialized version that `bar` is the underlying
+type we said it was in the `typedef`. This is unlike the default in this RFC
+but the same as when we use `default { .. }`.
+
+Do note however that C++ will allow us to remove `make_a_bar` from the
+specialized `foo<wrap<T>>` but will then error out with (gcc 4.6.2):
+
+```cc
+main.cpp: In function 'int main()':
+main.cpp:21:24: error: 'struct foo<wrap<void> >' has no member named 'make_a_bar'
+     std::cout << b_foo.make_a_bar() << std::endl;
+                        ^~~~~~~~~~
+```
+
+This shows that templates in C++ are fundamentally different from the type of
+parametric polymorphism (generics) that Rust employs.
```

Comment:
I would cut out this part, from the "do note" on, and rewrite the above paragraph.